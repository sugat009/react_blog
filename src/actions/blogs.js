import uuid from 'uuid';
import { database } from '../firebase/firebase';


// add blog post
export const addBlog = (blog) => ({
    type: 'ADD_BLOG',
    blog
});

// start add blog (async action for addBlog)
export const startAddBlog = (blogData = {}) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        const {
            title = '',
            content = '',
            createdAt = 0
        } = blogData;
        const blog = { title, content, createdAt };
        return database.ref(`users/${uid}/blogs`).push(blog)
            .then((ref) => {
                dispatch(addBlog({
                    id: ref.key,
                    ...blog
                }));
            });
    };
};

// remove blog post
export const removeBlog = ({ id } = {}) => ({
    type: 'REMOVE_BLOG',
    id
});

export const startRemoveBlog = ({ id } = {}) => {
    return (dispatch, getState) => {
        return database.ref(`users/${getState().auth.uid}/blogs/${id}`).remove().then(() => {
            dispatch(removeBlog({ id }));
        });
    };
};


// edit blog post
export const editBlog = (id, updates) => ({
    type: 'EDIT_BLOG',
    id,
    updates
});

export const startEditBlog = (id, updates) => {
    return (dispatch, getState) => {
        return database.ref(`users/${getState().auth.uid}/blogs/${id}`).update(updates)
            .then(() => {
                dispatch(editBlog(id, updates));
            });
    };
};


// SET Blogs part
export const setBlog = (blogs) => ({
    type: 'SET_BLOG',
    blogs
});

export const setStartBlog = () => {
    return (dispatch, getState) => {
        return database.ref(`users/${getState().auth.uid}/blogs`).once('value').then((snapshot) => {
            const blogData = [];
            snapshot.forEach((childSnapshot) => {
                blogData.push({
                    id: childSnapshot.key,
                    ...childSnapshot.val()
                });
            });
            dispatch(setBlog(blogData));
        });
    };  
};

// GET blogs part
export const getBlog = (blogs) => ({
    type: 'GET_BLOG',
    blogs
});

export const getStartBlog = () => {
    return (dispatch, getState) => {
        return database.ref(`users`).once('value').then((snapshot) => {
            const blogData = [];
            snapshot.forEach((childSnapshot) => {
                if (childSnapshot.key !== getState().auth.uid) {
                    blogData.push({
                        id: childSnapshot.key,
                        ...childSnapshot.val().blogs
                    });
                }
            });
            dispatch(getBlog(blogData));
            console.log(blogData);
        });
    };
};