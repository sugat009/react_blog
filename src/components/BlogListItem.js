import React from 'react';
import { Link } from 'react-router-dom';
import moment from 'moment';

const BlogListItem = ({ id, title, content, createdAt, user }) => (
    <Link to={`/edit/${id}`} className="list-item">
        <div>
            <h3 className="list-item__title">{title}</h3>
            <span className="list-item__subtitle">{moment(createdAt).format("MMMM Do, YYYY")}</span>
        </div>
        <h3 className="list-item__data">{user.displayName}</h3>
    </Link>
);

export default BlogListItem;