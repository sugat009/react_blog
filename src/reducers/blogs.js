const blogReducerDefaultState =  [];

export default (state = blogReducerDefaultState, action) => {
    switch (action.type) {
        case 'ADD_BLOG':
            return [
                ...state,
                action.blog
            ];
        case 'REMOVE_BLOG':
            return state.filter(({ id }) => id !== action.id);
        case 'EDIT_BLOG':
            return state.map((blog) => {
                if (blog.id === action.id) {
                    return {
                        ...blog,
                        ...action.updates
                    }
                } else {
                    return blog;
                };
            });
        case 'SET_BLOG':
            console.log(action.blogs, "aaaaaaaaaaaaaaaaaaaa");
            return action.blogs;
        case 'GET_BLOG':
            console.log(action.blogs);
            return action.blogs;
        default:
            return state;
    }
};