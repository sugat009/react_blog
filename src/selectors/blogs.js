import moment from 'moment';

// get visible blogs after the filters have been applied

export default (blogs, { title, sortBy, startDate, endDate }) => {
    return blogs.filter((blog) => {
        const createdAtMoment = moment(blog.createdAt);
        const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true;
        const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true;
        const titleMatch = blog.title.toLowerCase().includes(title.toLowerCase());
        
        return startDateMatch && endDateMatch && titleMatch;
    }).sort((a, b) => {
        if (sortBy === 'date') {
            return a.createdAt < b.createdAt ? 1 : -1;
        } else if (sortBy === 'alphabet') {
            return a.title > b.title ? 1 : -1;
        }
    });
};