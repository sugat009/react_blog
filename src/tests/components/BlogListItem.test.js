import React from 'react';
import { shallow } from 'enzyme';
import blogs from '../fixtures/blogs';
import BlogListItem from '../../components/BlogListItem';


test('should render BlogListItem correctly', () => {
    const wrapper = shallow(<BlogListItem {...blogs[1]} />);
    expect(wrapper).toMatchSnapshot();
});