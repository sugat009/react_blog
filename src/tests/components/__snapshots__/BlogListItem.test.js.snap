// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render BlogListItem correctly 1`] = `
<div>
  <Link
    replace={false}
    to="/edit/2"
  >
    <h3>
      ethical hacking
    </h3>
    <p>
      fraand ill show u some ethical hacking
       - 
      December 27th, 1969
    </p>
  </Link>
</div>
`;
